lib/wasix/src/runtime/module_cache/filesystem.rs

```diff
@@ -37,45 +40,57 @@ impl FileSystemCache {
 
 #[async_trait::async_trait]
 impl ModuleCache for FileSystemCache {
-    #[tracing::instrument(level = "debug", skip_all, fields(%key))]
+    #[tracing::instrument(level = "debug", skip_all, fields(% key))]
     async fn load(&self, key: ModuleHash, engine: &Engine) -> Result<Module, CacheError> {
         let path = self.path(key, engine.deterministic_id());
 
-        // FIXME: This will all block the thread at the moment. Ideally,
-        // deserializing and uncompressing would happen on a thread pool in the
-        // background.
-        // https://github.com/wasmerio/wasmer/issues/3851
-
-        let bytes = read_file(&path)?;
-
-        match deserialize(&bytes, engine) {
-            Ok(m) => {
-                tracing::debug!("Cache hit!");
-                Ok(m)
-            }
-            Err(e) => {
-                tracing::debug!(
-                    %key,
-                    path=%path.display(),
-                    error=&e as &dyn std::error::Error,
-                    "Deleting the cache file because the artifact couldn't be deserialized",
-                );
-
-                if let Err(e) = std::fs::remove_file(&path) {
-                    tracing::warn!(
-                        %key,
-                        path=%path.display(),
-                        error=&e as &dyn std::error::Error,
-                        "Unable to remove the corrupted cache file",
-                    );
+        self.task_manager
+            .runtime_handle()
+            .spawn({
+                let task_manager = self.task_manager.clone();
+                let engine = engine.clone();
+
+                async move {
+                    let bytes = read_file(&path).await?;
+
+                    task_manager
+                    .spawn_await({
+
+                        move || match deserialize(&bytes, &engine) {
+                            Ok(m) => {
+                                tracing::debug!("Cache hit!");
+                                Ok(m)
+                            }
+                            Err(e) => {
+                                tracing::debug!(
+                            %key,
+                            path=%path.display(),
+                            error=&e as &dyn std::error::Error,
+                            "Deleting the cache file because the artifact couldn't be deserialized",
+                        );
+
+                                if let Err(e) = std::fs::remove_file(&path) {
```

Comment:
Should this be done using a `tokio` function?

Reply:
Since the code is run on one of `rusty_pool` threads, it's okay.